crates/storage/provider/src/post_state.rs

```diff
@@ -338,66 +272,94 @@ impl PostState {
     /// 1. This post-state has the changes reverted
     /// 2. The returned post-state does *not* have the changes reverted, but only contains the
     /// descriptions of the changes that were reverted in the first post-state.
-    pub fn split_at(&mut self, transition_id: TransitionId) -> Self {
+    pub fn split_at(&mut self, revert_to_block: BlockNumber) -> Self {
         // Clone ourselves
         let mut non_reverted_state = self.clone();
 
         // Revert the desired changes
-        let reverted_changes = self.revert_to(transition_id);
+        self.revert_to(revert_to_block);
 
-        // Compute the new `current_transition_id` for `non_reverted_state`.
-        let new_transition_id =
-            reverted_changes.last().map(|c| c.transition_id()).unwrap_or_default();
-        non_reverted_state.changes = reverted_changes;
-        non_reverted_state.current_transition_id = new_transition_id + 1;
+        // Remove all changes in the returned post-state that were not reverted
+        non_reverted_state
+            .storage_changes
+            .retain(|block_number, _| *block_number > revert_to_block);
+        non_reverted_state
+            .account_changes
+            .retain(|block_number, _| *block_number > revert_to_block);
 
         non_reverted_state
     }
 
     /// Add a newly created account to the post-state.
-    pub fn create_account(&mut self, address: Address, account: Account) {
-        self.add_and_apply(Change::AccountCreated {
-            id: self.current_transition_id,
-            address,
-            account,
-        });
+    pub fn create_account(
+        &mut self,
+        block_number: BlockNumber,
+        address: Address,
+        account: Account,
+    ) {
+        self.accounts.insert(address, Some(account));
+        self.account_changes.entry(block_number).or_default().entry(address).or_insert(None);
     }
 
     /// Add a changed account to the post-state.
     ///
     /// If the account also has changed storage values, [PostState::change_storage] should also be
     /// called.
-    pub fn change_account(&mut self, address: Address, old: Account, new: Account) {
-        self.add_and_apply(Change::AccountChanged {
-            id: self.current_transition_id,
-            address,
-            old,
-            new,
-        });
+    pub fn change_account(
+        &mut self,
+        block_number: BlockNumber,
+        address: Address,
+        old: Account,
+        new: Account,
+    ) {
+        self.accounts.insert(address, Some(new));
+        self.account_changes.entry(block_number).or_default().entry(address).or_insert(Some(old));
     }
```

Comment:
@rakita you mentioned that this is incorrect right now? 

`or_insert` will keep the earliest state for the account, and only insert the value if it didn't exist, so that the block-level changeset gives us the pre-state. that seems correct? or am i misunderstanding?

Reply:
This part was correct, I found a different but similar bug for storage, and shared the link to this one as I thought it is the same problem.